### ISSUE TITLE
Evaluate Objective function Abstract Pyomo model

### ISSUE THREAD

**Issue by asuncionjc** [Pyomo/pyomo#1148]
Hi all,
my name is Asunción Jiménez-Cordero. I am a mathematical (machine learning) researcher from the University of Málaga (Spain). I recently begin to use Python and Pyomo to solve optimization problems.

I wonder if it is possible to evaluate the objective function for particular values in the variables. More specifically, I am interested in knowing the value of the objective function for the initial values of decision variables.

Below, you can find an outline of what my script should be:

```
- Read the abstract model.
- Create an instance.
- Set the parameters for the abstract model.
- Initialize the decision variables for the abstract model.
- Evaluate the objective function (and luckily also the constraints) at the decision variables previously defined.
```
A minimum working example (MWE) of what I want to implement is given in the following lines:
```python
import pyomo.environ as pe

# The definition of the model is not important. A simple model is then created:

def my_model_definition():

  my_model = AbstractModel()

  my_model.indexes = pe.RangeSet(10)
  my_model.x_variables = pe.Var(my_model.indexes)
  my_model.sample_parameters = pe.Param(my_model.indexes)
  
  my_model.objective_function = pe.Objective(rule = my_objective_function,
                                              sense = pe.minimize)
  my_model.constraint = pe.Constraint(rule= my_constraint)

  return my_model

#This is very simple objective function:

def my_objective_function(my_model):
  objective_value = sum(my_model.x_variables[index]*my_model.sample_parameters[index] for index in model.indexes)
  return objective_value

# Let us formulate a toy example constraint:

def my_constraint(my_model):
    constraint_value = my_model.x_variables[1] - my_model.x_variables[2] == 0
    return constraint_value

```
After the model definition, I would like to execute the following lines. I indicate in a comment the step I am executing, according to the outline I write you above:

```python

# Read the abstract model.
my_model = my_model_definition()

# Create an instance.
my_instance = my_model.create_instance()

# Set the parameters for the abstract model (and define the initial solution for the variables in the same loop).
dictionary_initial_variables = {}
for index in range(1, 10):
    my_instance.sample_parameters[index] = index
    dictionary_initial_variables[index] = 10*index

# Initialize the decision variables for the abstract model.
my_instance.x_variables.initialize = dictionary_initial_variables

#  Evaluate the objective function (and luckily also the constraints) at the decision variables previously defined.
# THIS IS THE KEY POINT.  
````
How can I know the value of the objective function for the values given by "dictionary_initial_variables"?
 I guess that should be something like:

```python
 my_instance.my_objective_function(x_variables = dictionary_initial_variables)
```
or something similar. However, any of the choices I tested returns in an error. 

Could any of you please help me with this problem?

If further information is necessary, do not hesitate to ask me.

Thank you in advance 😊😊😊

**Comment by carldlaird** [comment]
If you initialize the variables with the desired values, you can ask for the value of the objective function with the pyomo value function.

e.g.,

print(pe.value(my_model.objective_function))

**Comment by asuncionjc** [comment]
Thanks for your comment @carldlaird 

Unfortunately, it does not work. 😥The following error message is returned (I have adapted the variable names for the MWE):

```
ERROR: evaluating object as numeric value: x_variables[1]
        (object: <class 'pyomo.core.base.var._GeneralVarData'>)
    No value for uninitialized NumericValue object x_variables[1]
ERROR: evaluating object as numeric value: my_objective_function
        (object: <class 'pyomo.core.base.objective.SimpleObjective'>)
    No value for uninitialized NumericValue object x_variables[1]
*** ValueError: No value for uninitialized NumericValue object x_variables[1]

```

It seems as if Pyomo was not able to evaluate the objective function, because the x_variables are not well defined. However, I initialized them as indicated in my previous comment, i.e.,

```python
my_instance.x_variables.initialize = dictionary_initial_variables
```
where `dictionary_initial_variables` is a dictionary that I have defined as follows:

```python
dictionary_initial_variables = {}
for index in range(1, 10):
    dictionary_initial_variables[index] = 10*index
```
Can anyone tell me if I am correctly initializing the variables? Does there exist an alternative?

Thank you all!

**Comment by carldlaird** [comment]
I have never done this by setting the initialize attribute. I suspect that is only fired when the model is first created and not afterwards. I usually do this with concrete models in the rule for the variables themselves or by directly setting the value.

I would try the following:

for k in my_instance.x_variables:
                my_instance.x_variables[k].value = dictionary_initial_values[k]

**Comment by DLWoodruff** [comment]
Given the way you are working,  you might be happier using a ConcreteModel.
If you want to keep using an AbstractModel, then I recommend that you give
data to the model when you instantiate it
(in the create_instance call; see
https://pyomo.readthedocs.io/en/stable/working_abstractmodels/data/raw_dicts.html#using-a-python-dictionary
and
https://pyomo.readthedocs.io/en/stable/working_abstractmodels/instantiating_models.html#instantiating-models
)

After your model is instantiated with data, you can evaluate the value of
Pyomo
objects using the Value function (there are many examples of the use of the
value function here:
https://pyomo.readthedocs.io/en/stable/working_models.html#)

Dave

**Comment by asuncionjc** [comment]
Thank you @carldlaird and @DLWoodruff for your answers 😄

@carldlaird : I tried to initialize the variables as you propose, instead of using the `initialize` attribute and now I can get the objective value 👏 It seems that the `initialize` attribute only works for Concrete models as you suggested, although Pyomo returns no error when using it in an Abstract model.

@DLWoodruff : I agree with you that I become happier if I use Concrete models. However, I plan to parallelize the code using the appropriate Pyomo and Python libraries, and as far I have understood, the parallelization with Pyomo and Python should be done with Abstract models 🤷‍♀️
Moreover, I find very useful the links you recommended to me regarding the instantiate of the models. Unfortunately, I cannot use it to initialize the parameters, because the parameters in my model are defined as `mutable`. Nevertheless, I will use them in the future when no mutable parameters are considered.

Now, my problem has been solved, and therefore I will close this issue. 

Thank you very much once again 🎉
